kv/kv.go

```diff
@@ -335,13 +337,148 @@ func (t StoreType) Name() string {
+type KeyRanges struct {
+	ranges [][]KeyRange
+
+	isPartitioned bool
+}
+
+// NewPartitionedKeyRanges constructs a new RequestRange for partitioned table.
+func NewPartitionedKeyRanges(ranges [][]KeyRange) *KeyRanges {
+	return &KeyRanges{
+		ranges:        ranges,
+		isPartitioned: true,
+	}
+}
+
+// NewNonParitionedKeyRanges constructs a new RequestRange for a non partitioned table.
+func NewNonParitionedKeyRanges(ranges []KeyRange) *KeyRanges {
+	return &KeyRanges{
+		ranges:        [][]KeyRange{ranges},
+		isPartitioned: false,
+	}
+}
+
+// FirstPartitionRange returns the the result of first range.
+// We may use some func to generate ranges for both partitioned table and non partitioned table.
+// This method provides a way to fallback to non-partitioned ranges.
+func (rr *KeyRanges) FirstPartitionRange() []KeyRange {
+	if len(rr.ranges) == 0 {
+		return []KeyRange{}
+	}
+	return rr.ranges[0]
+}
+
+// SetToNonPartitioned set the status to non-partitioned.
+func (rr *KeyRanges) SetToNonPartitioned() error {
+	if len(rr.ranges) > 1 {
+		return errors.Errorf("you want to change the partitioned ranges to non-partitioned ranges")
+	}
+	rr.isPartitioned = false
+	return nil
+}
+
+// AppendSelfTo appends itself to another slice.
+func (rr *KeyRanges) AppendSelfTo(ranges []KeyRange) []KeyRange {
+	for _, r := range rr.ranges {
+		ranges = append(ranges, r...)
+	}
+	return ranges
+}
+
+// SortByFunc sorts each partition's ranges.
+// Since the ranges are sorted in most cases, we check it first.
+func (rr *KeyRanges) SortByFunc(sortFunc func(i, j KeyRange) bool) {
+	if !slices.IsSortedFunc(rr.ranges, func(i, j []KeyRange) bool {
+		// A simple short-circuit since the empty range actually won't make anything wrong.
+		if len(i) == 0 || len(j) == 0 {
+			return true
+		}
+		return sortFunc(i[0], j[0])
+	}) {
+		slices.SortFunc(rr.ranges, func(i, j []KeyRange) bool {
+			if len(i) == 0 {
+				return true
+			}
+			if len(j) == 0 {
+				return false
+			}
+			return sortFunc(i[0], j[0])
+		})
+	}
+	for i := range rr.ranges {
+		if !slices.IsSortedFunc(rr.ranges[i], sortFunc) {
+			slices.SortFunc(rr.ranges[i], sortFunc)
+		}
+	}
+}
+
+// ForEachPartitionWithErr runs the func for each partition with an error check.
+func (rr *KeyRanges) ForEachPartitionWithErr(theFunc func([]KeyRange) error) (err error) {
+	for i := range rr.ranges {
+		err = theFunc(rr.ranges[i])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// ForEachPartition runs the func for each partition without error check.
+func (rr *KeyRanges) ForEachPartition(theFunc func([]KeyRange)) {
+	for i := range rr.ranges {
+		theFunc(rr.ranges[i])
+	}
+}
+
+// PartitionNum returns how many partition is involved in the ranges.
+func (rr *KeyRanges) PartitionNum() int {
+	return len(rr.ranges)
+}
+
+// IsFullySorted checks whether the ranges are sorted inside partition and each partition is also sorated.
+func (rr *KeyRanges) IsFullySorted() bool {
```

Comment:
we should add some comments about:
order property being kept by the range partitioned table only.